### PR TITLE
feat(training): enable MFSDP V2

### PR DESCRIPTION
**Pull request by wujingyue**
## What

- Bump Megatron-LM to `59b72fa57` and regenerate `uv.lock` for MFSDP V2 support.
- Keep MFSDP V1 as the default while making `ddp.megatron_fsdp_version=2` opt into the experimental V2 path.
- Enforce V2's supported DP-only BF16 configuration and reject checkpointing, gradient clipping, and other unsupported features early.
- Add a two-GPU, ten-step dense `HybridModelProvider` smoke test using `hybrid_layer_pattern="**"` and `clip_grad=0.0`.
- Remove unused temporary-directory fixtures from the non-checkpoint MFSDP tests; retain the checkpoint-resume test's real temporary directory.

## Why

MFSDP V2 has a distinct wrapper and optimizer contract from V1. Bridge must select its safe configuration explicitly and avoid V1-only assumptions.

## Validation

- `uv run --no-sync pre-commit run --all-files`
- Two-GPU functional smoke: existing MFSDP V1 basic training and the new MFSDP V2 dense-Hybrid test.

Related to NVIDIA/Megatron-LM#6325.

### CODE REVIEW

File: 3rdparty/Megatron-LM
Comment (wujingyue) [inline 3rdparty/Megatron-LM:1]: This PR requires to sync past https://github.com/NVIDIA/Megatron-LM/pull/5865. Should I wait for the next time you bump up the Megatron-LM version?

File: uv.lock
Comment (wujingyue) [inline uv.lock:1]: Not sure where this change came from. Probably an artifact from updating 3rdparty/Megatron-LM.

**Comment by wujingyue** [comment]
cc @rapatel

File: 3rdparty/Megatron-LM
Comment (cuichenx) [reply 3rdparty/Megatron-LM:1]: can this wait until #5431 is merged? sorry this bump is taking a while

Comment (wujingyue) [reply 3rdparty/Megatron-LM:1]: No problem at all!

File: tests/functional_tests/test_groups/megatron_fsdp/test_megatron_fsdp.py [hunk 1]
```diff
@@ -479,7 +540,7 @@ def test_fsdp_pretrain_save_resume(self, tmp_path):
             clear_directories(shared_base_dir)
 
     @pytest.mark.run_only_on("GPU")
-    def test_fsdp_precision_aware_optimizer_decoupled_grad_consistency(self, tmp_path):
```
Comment (wujingyue) [inline tests/functional_tests/test_groups/megatron_fsdp/test_megatron_fsdp.py:482]: I removed tmp_path because it's not used by the test. As a result, I also removed the try-finally.

File: 3rdparty/Megatron-LM
Comment (wujingyue) [reply 3rdparty/Megatron-LM:1]: Hey @cuichenx and @yaoyu-33, I rebased this PR to main. Will you review this?

**Review by cuichenx — COMMENTED** [review]
i would suggest introducing a fsdp_compat.py or mcore_compat.py file, and then do the try except FSDP import and MEGATRON_FSDP_TYPES in that one place instead of everywhere. Do you think that makes sense? Otherwise LGTM

**Comment by wujingyue** [comment]
> i would suggest introducing a fsdp_compat.py or mcore_compat.py file, and then do the try except FSDP import and MEGATRON_FSDP_TYPES in that one place instead of everywhere. Do you think that makes sense? Otherwise LGTM

Makes sense -- I've made that change in the latest commit. I'm not sure about the new dependency from src/megatron/bridge/models/conversion/utils.py to src/megatron/bridge/training/; let me know if you see a better approach. cc @cuichenx

**Review by cuichenx — APPROVED** [review]
LGTM
